### ISSUE TITLE
keras model.compile: allow user-specific label shapes/dtypes for loss/metrics

### ISSUE THREAD

**Issue by jackd** [tensorflow/tensorflow#25957]
## System information
- TensorFlow version: 1.12 (also present in 1.13, 2.0)
## Are you willing to contribute it (Yes/No):
Yes

## Describe the feature and the current behavior/state.
`tf.keras.models.Model.compile` currently has no way of knowing the shape/size of targets that will be passed to it during `Model.fit` (and other main looping methods). Instead it creates placeholders on a 1-to-1 basis corresponding to the same dtype and number of dimensions as each output. This is a safe assumption for most regression tasks, but leads to redundant casting/reshaping for even simple classification problems. More importantly, it leads to very confusing behaviour, loss/metric functions which need to cater for a wider variety of inputs than necessary and an inability to do proper error checking.

For example, in 2.0 `tf.keras.backend.sparse_categorical_crossentropy`'s documentation claims the first argument is `target: An integer tensor.` However, there's absolutely nothing stopping you from providing float values. This is because the implementation needs to be able to accept arbitrarily sized float values when passed placeholders during `compile`, then simply flattens and casts the values (which has no affect on inputs already of the correct shape/type).

This is particularly confusing for users trying to implement their own loss function. Mimicing the advertised interface will result in errors.

This isn't just a documentation issue however: if inappropriate float values are passed in (i.e. values that aren't the result of casting integers to floats) this will be very difficult to pick up.

If users are asked to provide `loss` and `metrics` functions, they will generally know the shape and types of the labels. Allowing them to specify this so their loss/metric functions only need to cater for these removes unecessary reshaping and casting.

## Will this change the current api? How?
I don't understand why loss/metric functions are called during compile with placeholders rather than being lazily evaluated during `fit`. From my perspective that would be the thing to change, resulting in no changes to the API. However, it may result in too much of a breaking change in other niche cases.

Lacking that, adding extra kwargs to `Model.compile - maybe `target_shapes`, `target_sizes` (`targets` is already taken) would be a non-breaking change. Default values can be created using the current assumptions.

## Who will benefit with this feature?
- People writing custom losses/metrics with the expectation that `y_true` shape/dtype will match that of the values in their labels.
- Anyone training with models using unecessary casts/reshapes (probably not a major performance issue).

**Comment by rmothukuru** [comment]
@jackd,
Sorry for the delayed response. 

**`Shape`** and **`Data Types`** of **`Loss`** are clearly specified in the [Documentation of Model.Compile](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile).

Also, the documentation of [Sparse Categorical Cross Entropy](https://www.tensorflow.org/api_docs/python/tf/keras/losses/sparse_categorical_crossentropy#args) no longer states that **`First Argument should be an Integer Tensor`**.

Thanks!
